### DISCUSSION TITLE
Why does JobDataMap not support generic functions?

### DISCUSSION THREAD

**Discussion by scotly** [quartznet/quartznet#2296] [Q&A] [answered]
now,JobDataMap is IDictionary<string,object>, I often need to serialize object , and  deserialize when fetching data.
so I suggest maintaining a IList<NameValue<string,TValue>> data structure in JobData 
no box when  UsingJobData, and no box when JobDataMap.Get<TValue> ,and it's convenient

**Comment by lahma** [comment]
Map is a common idiom to make dictionary-based lookups. I don't see how `TValue` would allow `string` and `int` as values for example.

**Comment by scotly** [comment]
**This change is likely to be significant!** 
I think it like this:

1. define an interface for JobData
```
public interface IJobData<TKey>
{
    /// <summary>
    /// Key.
    /// </summary>
    TKey Key { get; }

}
```

2.  Implement the interface
```
public class JobData<TKey, TValue> : IJobData<TKey>
{
    /// <summary>
    /// Key.
    /// </summary>
    public TKey Key{ get; set; }

    /// <summary>
    /// Value.
    /// </summary>
    public TValue Value { get; set; }

    public JobData(TKey name, TValue value)
    {
        Key = name;
        Value = value;
    }
}
```
3. next,define a collection for IJobData
```
public class JobDataCollection<TKey> where TKey : notnull
{
    public List<IJobData<TKey>> JobDataList { get; private set; }
    public JobDataCollection()
    {
        JobDataList = new List<IJobData<TKey>>();
    }

    public bool IsEmpty => JobDataList == null || !JobDataList.Any();

    /// <summary>
    /// add JobData
    /// </summary>
    /// <typeparam name="TValue"></typeparam>
    /// <param name="key"></param>
    /// <param name="value"></param>
    /// <returns></returns>
    public JobData<TKey, TValue> Add<TValue>(TKey key, TValue value)
    {
        var jobData = new JobData<TKey, TValue>(key, value);
        JobDataList.Add(jobData);

        return jobData;
    }

    /// <summary>
    /// Get JobData
    /// </summary>
    /// <typeparam name="TValue"></typeparam>
    /// <param name="key"></param>
    /// <returns></returns>
    public JobData<TKey, TValue> GetJobdata<TValue>(TKey key)
    {
        try
        {
            return (JobData<TKey, TValue>) JobDataList.FirstOrDefault(k => k.Key.Equals(key))!;
        }
        catch
        {
            //TODO Log
        }
        return null!;
    }

    /// <summary>
    /// Remove
    /// </summary>
    /// <param name="key"></param>
    public void Remove(TKey key)
    {
        JobDataList.RemoveAll(k => k.Key.Equals(key));
    }

    /// <summary>
    /// Clear All
    /// </summary>
    public void Clear() => JobDataList.Clear();
}
```

4. Implement a string JobDataMap
```
public class StringJobDataMap
{
    private JobDataCollection<string> JobDataCollection;

    public bool IsEmpty => JobDataCollection.IsEmpty;

    public StringJobDataMap()
    {
        JobDataCollection = new JobDataCollection<string>();
    }

    public JobData<string, TValue> Put<TValue>(string key, TValue value)
    {
        return JobDataCollection.Add(key, value);
    }

    public void Put(IJobData<string> data) => JobDataCollection.JobDataList.Add(data);

    public void PutAll(StringJobDataMap jobDataMap)
    {
        JobDataCollection.Clear();
        foreach (var item in jobDataMap.JobDataCollection.JobDataList)
        {
            this.Put(item);
        }
    }

    public JobData<string, TValue> GetValue<TValue>(string key) => JobDataCollection.GetJobdata<TValue>(key);

    public StringJobDataMap Clone()
    {
        var map = new StringJobDataMap();
        map.PutAll(this);

        return map;
    }
}
```

5.Replace jobdatamap with StringJobDataMap

Do you think it's feasible?

**Comment by lahma** [comment]
I probably don't fully understand the concept you are proposing, what I see is odd dictionary allowing multiple keys and also impossible to have working in a backwards compatible manner?

**Reply by scotly** [reply]
I don't quite understand why you have to use a dictionary?What I mean by this is to replace dictionaries with List generics,
backwards compatibility is taken into account, and it is quick to find,So the Key is defined。
If you want to keep the key unique like a dictionary，When Put, remove the jobdata of the existing key,then add new item.
Actually, what I meant was that Key was dispensable,JobDataMap is custom data，It is up to the developer to decide whether the content of the collection can be duplicated,We just need to return a collection of the same key

**Comment by TomLint** [comment]
You could just write yourself a set of extension methods to convert an object into a dictionary and vice versa, if you really want to do it this way, but why not simply store this object in the data map as a value?

**Comment by lahma** [comment, marked as answer]
It does now, in 4.0: `map.Get<T>(key)`, `map.TryGet<T>(key, out T value)` and
`map.GetValueOrDefault<T>(key, fallback)`, beside the per-type accessors. `Get<T>` throws
`KeyNotFoundException` or `InvalidCastException` naming the key; `GetValueOrDefault<T>` takes your
fallback; all three are a pure type test, no string parsing. They are extension members, so they work on
`JobDataMap` and on `SchedulerContext` alike. Shipped in 4.0.0-alpha.2 (#3396).

What did not change, and will not, is the storage: the map stays `IDictionary<string, object?>`. A
`List<IJobData<TKey>>` still erases to `object` the moment it has to hold a `string` and an `int`
together, so the boxing is inherent in a heterogeneous map — the list would cost the O(1) lookup and the
persisted format every ADO store already writes, and buy nothing back.

@TomLint is right that storing the object directly is the answer. Read what that commits you to first —
the serializer has to keep reading it after you rename a property:
[The Job Data Map](https://www.quartz-scheduler.net/documentation/quartz-4.x/tutorial/job-data-map.html#the-read-side-typed-accessors).

Try `Get<T>` on 4.0.0-alpha.2 and say whether it covers what you were after.
